### PR TITLE
Tighten up Dask serialization checks

### PR DESCRIPTION
**Pull request by jakirkham**
Fixes https://github.com/rapidsai/cudf/issues/4421

Adds a few checks around serialization. Some of these are safe-guards. Some of these will raise errors to catch problematic behavior we have seen. Hopefully this will make sure we are getting top-performance when serializing. Also should make it easier to catch the cases where we are not.

cc @cjnolet (who raised this point offline)
cc @dantegd (for awareness)

### CODE REVIEW

**Comment by jakirkham** [comment]
Sorry for all the noise. Still learning how to use labels 😅

**Comment by jakirkham** [comment]
Added PR ( https://github.com/rapidsai/cuml/pull/1848 ) downstream to ensure these are applied to `CumlArray` as well.

**Comment by jakirkham** [comment]
Also tests appear to be failing due to package checksum errors?

https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fcudf%2Fprb%2Fcudf-cpu-build/detail/cudf-cpu-build/36547/pipeline

**Comment by kkraus14** [comment]
rerun tests

**Comment by jakirkham** [comment]
Thanks Keith! 😀

File: python/cudf/cudf/comm/serialize.py [hunk 1]
```diff
@@ -23,22 +23,29 @@
...
 
     @cuda_deserialize.register(serializable_classes)
     @dask_deserialize.register(serializable_classes)
     def deserialize_cudf_object(header, frames):
         with log_errors():
+            if header["serializer"] == "cuda":
+                assert all(
+                    hasattr(f, "__cuda_array_interface__") for f in frames
+                )
+            if header["serializer"] == "dask":
```
Comment (jakirkham) [inline python/cudf/cudf/comm/serialize.py]: Should have been `elif`. Fixing in PR ( https://github.com/rapidsai/cudf/pull/4537 ).

Comment (kkraus14) [reply python/cudf/cudf/comm/serialize.py]: How can the key `serializer` have multiple values?

Comment (quasiben) [reply python/cudf/cudf/comm/serialize.py]: Often in dask a list of possible serializers is passed:
> serializers=("cuda", "dask", "pickle", "error")

if one fails the next one is used

Comment (jakirkham) [reply python/cudf/cudf/comm/serialize.py]: Yeah though the `header` should only contain one at the top-level. IOW this shouldn't be a collection.

Comment (jakirkham) [reply python/cudf/cudf/comm/serialize.py]: This thread continues here: https://github.com/rapidsai/cudf/pull/4537#pullrequestreview-375722736

**Pull request merged**